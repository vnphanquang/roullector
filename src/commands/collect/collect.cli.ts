import {
  Command,
  InvalidArgumentError,
  Option,
} from 'commander';

import { collect } from '$commands/collect/collect';
import {
  keyTransformCliToFunc,
  defaultCollectOptions,
  defaultKeyTransformCli,
} from '$commands/collect/collect.constants';
import type {
  CollectOptions,
  KeyTransformFn,
} from '$commands/collect/collect.types';
import { KeyTransformCli } from '$commands/collect/collect.types';
import {
  commaSeparatedList,
  number,
} from '$utils/parser';

export function collectCli() {
  return new Command('collect')
    .description('collect route data')
    .option(
      '-i, --inDir <path>',
      'directory path to collect',
      defaultCollectOptions.inDir
    )
    .option(
      '-e, --extensions <list>',
      'comma-separated list of file extensions to collect',
      commaSeparatedList,
      defaultCollectOptions.extensions,
    )
    .option(
      '-x, --ignorePatterns <list>',
      'comma-separated list of filename patterns (string literal or regex) to ignore',
      commaSeparatedList,
      defaultCollectOptions.ignorePatterns,
    )
    .option(
      '--no-output',
      'don\'t write to disk but print to console instead',
      defaultCollectOptions.output,
    )
    .option(
      '-o, --outDir <path>',
      'outputs generated files to this folder',
      defaultCollectOptions.outDir,
    )
    .option(
      '-d, --depth <number>',
      'depth of inDir to collect',
      number,
      defaultCollectOptions.depth,
    )
    .addOption(
      new Option( '-t, --keyTransform <variant>', 'how to transform route key')
        .default([])
        .argParser<KeyTransformFn[]>((value, previous) => {
          if (!(value in KeyTransformCli)) {
            throw new InvalidArgumentError(`Must be one of: ${Object.keys(KeyTransformCli).join(', ')}`);
          }
          const valued = keyTransformCliToFunc[KeyTransformCli[value] as KeyTransformCli];
          return [...previous, valued];
        }),
    )
    .option(
      '-k, --dirkey <name>',
      'how to transform route key',
      defaultCollectOptions.dirkey,
    )
    .option(
      '--no-utils',
      'don\'t generate utils for building path with arguments',
      defaultCollectOptions.utils,
    )
    .option(
      '--no-typescript',
      'outputs files in javascript instead of typescript',
      defaultCollectOptions.typescript,
    )
    .option(
      '--verbose',
      'prints more info during operation',
      defaultCollectOptions.verbose,
    )
    .action((options: CollectOptions) => {
      if (options.keyTransform.length === 0) {
        options.keyTransform = defaultKeyTransformCli.map((k) => keyTransformCliToFunc[k]);
      }
      if (options.verbose) {
        console.info('Running "collect" with options:');
        console.table(Object.fromEntries(
          Object.entries(options).map(
            ([key, value]) => ([
              key,
              { value },
            ])
          )),
        );
      }

      const output = collect(options);
      if (!options.output) {
        console.log(output);
      }
    });
}
