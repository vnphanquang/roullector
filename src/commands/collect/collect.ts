import {
  readdirSync,
  mkdirSync,
  existsSync,
  rmdirSync,
  writeFileSync,
} from 'fs';
import { resolve } from 'path';

import type {
  CollectOptions,
  CollectOutput,
} from '$commands/collect/collect.types';
import { generateMapping } from '$commands/collect/generators/mapping';
import { generateRouteUtil } from '$commands/collect/generators/route';

import {
  defaultCollectOptions,
  MAPPING_FILENAME,
} from './collect.constants';

export function collect(options: Partial<CollectOptions> = {}): CollectOutput {
  const mergedOptions: CollectOptions = {
    ...defaultCollectOptions,
    ...options,
  };
  const { outDir, output, typescript } = options;
  const result = {
    json: null,
    route: null,
  };
  try {
    const { map } = generateMapping(mergedOptions);

    let json: CollectOutput['json'] = JSON.stringify(map, null, 2);
    let route: CollectOutput['route'] = mergedOptions ? generateRouteUtil(mergedOptions) : null;

    if (output) {
      if (!existsSync(outDir)) {
        mkdirSync(outDir, { recursive: true });
      }

      // write json
      let file = resolve(outDir, MAPPING_FILENAME);
      writeFileSync(file, json, { encoding: 'utf-8' });
      json = file;

      // write index
      const ext = typescript ? '.ts' : '.js';
      file = resolve(outDir, 'index' + ext);
      writeFileSync(file, route, { encoding: 'utf-8' });
      route = file;
    }

    result.json = json;
    result.route = route;
  } catch (error) {
    console.error(error);
    if (existsSync(outDir) && readdirSync(outDir).length === 0) {
      rmdirSync(outDir, { recursive: true });
    }
    throw error;
  }
  return result;
}
