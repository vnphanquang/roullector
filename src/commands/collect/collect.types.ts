export type CollectOptions = {
  /** input directory path to collect route data from */
  inDir: string;
  /** file extensions */
  extensions: string[];
  /** patterns to ignore files */
  ignorePatterns: (string|RegExp)[];
  /** whether to write to disk. If false will print to console in command-line mode, or returned in lib usage */
  output: boolean;
  /** route data output directory path */
  outDir: string;
  /** prints more info during operation */
  verbose: boolean;
  /** how to transform route key */
  keyTransform: KeyTransformFn[];
  /** depth of inDir to collect */
  depth: number;
  /** key to save path for directories with no index file */
  dirkey: string;
  /** generate utils for building path with arguments */
  utils: boolean;
  /** whether to output files in typescript */
  typescript: boolean;
};

export type CollectOutput = {
  /** path to generated json, if options.output is false, this is the generated output itself */
  json: null|string;
  /** path to generated route util, if options.output is false, this is the generated output itself */
  route: null|string;
};

export enum KeyTransformCli {
  none = 'none',
  camelCase = 'camelCase',
  dollarArg = 'dollarArg',
}

export type KeyTransformFn = (key: string, original: string) => string;
