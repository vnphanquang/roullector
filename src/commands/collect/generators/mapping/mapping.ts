import {
  statSync,
  readdirSync,
  existsSync,
} from 'fs';
import {
  basename,
  join,
} from 'path';

import type { CollectOptions } from '$commands/collect/collect.types';
import { compose } from '$utils/transformer';

export type DeepObject<T> = {
  [key: string]: T | DeepObject<T>;
};

export function generateMapping(options: CollectOptions, inDir: string = options.inDir, currentPath = '', level = 0): {
  map: DeepObject<string>;
  array: string[];
} | null {
  if (!existsSync(inDir)) return null;

  const children = readdirSync(inDir);

  const map: DeepObject<string> = {};
  const array: string[] = [];

  if (level <= options.depth) {
    for (const child of children) {
      const absPath = join(inDir, child);
      const stats = statSync(absPath);

      // skip if any ignore pattern matches
      const filename = basename(child);
      console.info(`Turbo ~ file: mapping.ts ~ line 36 ~ generateMapping ~ filename`, filename);
      if (options.ignorePatterns.some(pattern => new RegExp(pattern).test(filename))) continue;

      if (stats.isDirectory()) {
        const nextPath = `${currentPath}/${child}`;
        const { map: childMap, array: childArray } = generateMapping(options, absPath, nextPath, level + 1);
        if (Object.keys(childMap).length) {
          if (!childMap['index']) {
            childMap[options.dirkey] = nextPath;
          }
          const transformed = compose(child, ...options.keyTransform);
          map[transformed] = childMap;
          array.push(...childArray);
        }
      } else {
        // skip if extension doesn't match
        if (options.extensions.every((ext) => !filename.endsWith(ext))) continue;

        // transform filename to route key
        const routeKey = filename.replace(/\.[^/.]+$/, ''); // trim extension
        const transformed = compose(routeKey, ...options.keyTransform);

        let routeName = routeKey;
        if (routeKey === 'index') {
          // no explicit `index`
          routeName = '';
        }
        let nextPath = currentPath;
        if (routeName) {
          nextPath += `/${routeName}`;
        }

        if (!nextPath && level === 0) {
          // root level index route
          nextPath = '/';
        }

        map[transformed] = nextPath;
        array.push(nextPath);
      }
    }
  }

  return { map, array };
}
