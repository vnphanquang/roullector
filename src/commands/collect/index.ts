export * from './collect';
export * from './collect.types';
export { defaultCollectOptions } from '$commands/collect/collect.constants';
