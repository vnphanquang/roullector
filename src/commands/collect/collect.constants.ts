import type {
  CollectOptions,
  KeyTransformFn
} from '$commands/collect/collect.types';
import {KeyTransformCli} from '$commands/collect/collect.types';
import {
  camelCasify,
  dollarArgify,
} from '$utils/transformer';

export const MAPPING_FILENAME = 'routes.json';
export const UTIL_ROUTE_FILENAME = 'route';

export const defaultCollectOptions: CollectOptions = {
  inDir: 'src/routes',
  extensions: ['.svelte'],
  ignorePatterns: [/^_/],
  output: true,
  outDir: 'src/generated/routing',
  typescript: true,
  keyTransform: [dollarArgify, camelCasify],
  verbose: false,
  depth: Infinity,
  dirkey: '__dir',
  utils: true,
};

export const defaultKeyTransformCli: KeyTransformCli[] = [
  KeyTransformCli.camelCase,
  KeyTransformCli.dollarArg,
];

export const keyTransformCliToFunc: Record<KeyTransformCli, KeyTransformFn> = {
  none: (str) => str,
  camelCase: camelCasify,
  dollarArg: dollarArgify,
};
