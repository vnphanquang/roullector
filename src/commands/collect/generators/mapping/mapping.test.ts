import {
  vol,
  fs,
} from 'memfs';

import { defaultCollectOptions } from '$commands/collect/collect.constants';
import type { CollectOptions } from '$commands/collect/collect.types';
import { generateMapping } from '$commands/collect/generators/mapping';
import { FILES } from '$tests/fixtures/memfs';

jest.mock('fs', () => fs);
beforeEach(() => {
  vol.fromNestedJSON(FILES, defaultCollectOptions.inDir);
});
afterEach(() => {
  vol.reset();
});

const generateMappingWithOptions = (snapshotToMatch: string, options: Partial<CollectOptions> = {}) => {
  const generated = generateMapping({
    ...defaultCollectOptions,
    ...options,
  });
  expect(generated).toMatchSnapshot(snapshotToMatch);
};

test(`generate JSON but empty directory`, () => {
  vol.reset();
  const outputPath = generateMapping(defaultCollectOptions);
  expect(outputPath).toBe(null);
});

test(`generate mapping with default options`, () => {
  generateMappingWithOptions('generateMapping: default');
});

test(`generate mapping with a custom extension matcher`, () => {
  generateMappingWithOptions('generateMapping: custom extension matcher', { extensions: ['.ts'] });
});

test(`generate mapping with depth of 1`, () => {
  generateMappingWithOptions('generateMapping: depth 1', { depth: 1 });
});

const ignorePatterns = ['^_', '^posts'];
test(`generate mapping with custom ignorePatterns: ${ignorePatterns}`, () => {
  generateMappingWithOptions('generateMapping: custom ignorePatterns',{ ignorePatterns });
});

const dirkey = '__customDir__';
test(`generate Mapping with custom dirkey: ${dirkey}`, () => {
  generateMappingWithOptions('generateMapping: custom dirkey', { dirkey });
});
