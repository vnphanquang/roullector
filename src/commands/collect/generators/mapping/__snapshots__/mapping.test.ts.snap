// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate Mapping with custom dirkey: __customDir__: generateMapping: custom dirkey 1`] = `
Object {
  "array": Array [
    "/admin/users/[id]",
    "/admin/users/[id]/posts/[post_id]",
    "/admin/users/[id]/posts/l-[short-link]-l",
    "/admin/users/[id]/posts/s-[slug]",
    "/admin/users",
    "/",
    "/me",
    "/sign-in",
    "/sign-up",
  ],
  "map": Object {
    "admin": Object {
      "__customDir__": "/admin",
      "users": Object {
        "$id": Object {
          "index": "/admin/users/[id]",
          "posts": Object {
            "$postId": "/admin/users/[id]/posts/[post_id]",
            "__customDir__": "/admin/users/[id]/posts",
            "l$shortLink$L": "/admin/users/[id]/posts/l-[short-link]-l",
            "s$slug": "/admin/users/[id]/posts/s-[slug]",
          },
        },
        "index": "/admin/users",
      },
    },
    "index": "/",
    "me": "/me",
    "signIn": "/sign-in",
    "signUp": "/sign-up",
  },
}
`;

exports[`generate mapping with a custom extension matcher: generateMapping: custom extension matcher 1`] = `
Object {
  "array": Array [
    "/admin/users/[id]/types",
  ],
  "map": Object {
    "admin": Object {
      "__dir": "/admin",
      "users": Object {
        "$id": Object {
          "__dir": "/admin/users/[id]",
          "types": "/admin/users/[id]/types",
        },
        "__dir": "/admin/users",
      },
    },
  },
}
`;

exports[`generate mapping with custom ignorePatterns: ^_,^posts: generateMapping: custom ignorePatterns 1`] = `
Object {
  "array": Array [
    "/admin/users/[id]",
    "/admin/users",
    "/",
    "/me",
    "/sign-in",
    "/sign-up",
  ],
  "map": Object {
    "admin": Object {
      "__dir": "/admin",
      "users": Object {
        "$id": Object {
          "index": "/admin/users/[id]",
        },
        "index": "/admin/users",
      },
    },
    "index": "/",
    "me": "/me",
    "signIn": "/sign-in",
    "signUp": "/sign-up",
  },
}
`;

exports[`generate mapping with default options: generateMapping: default 1`] = `
Object {
  "array": Array [
    "/admin/users/[id]",
    "/admin/users/[id]/posts/[post_id]",
    "/admin/users/[id]/posts/l-[short-link]-l",
    "/admin/users/[id]/posts/s-[slug]",
    "/admin/users",
    "/",
    "/me",
    "/sign-in",
    "/sign-up",
  ],
  "map": Object {
    "admin": Object {
      "__dir": "/admin",
      "users": Object {
        "$id": Object {
          "index": "/admin/users/[id]",
          "posts": Object {
            "$postId": "/admin/users/[id]/posts/[post_id]",
            "__dir": "/admin/users/[id]/posts",
            "l$shortLink$L": "/admin/users/[id]/posts/l-[short-link]-l",
            "s$slug": "/admin/users/[id]/posts/s-[slug]",
          },
        },
        "index": "/admin/users",
      },
    },
    "index": "/",
    "me": "/me",
    "signIn": "/sign-in",
    "signUp": "/sign-up",
  },
}
`;

exports[`generate mapping with depth of 1: generateMapping: depth 1 1`] = `
Object {
  "array": Array [
    "/",
    "/me",
    "/sign-in",
    "/sign-up",
  ],
  "map": Object {
    "index": "/",
    "me": "/me",
    "signIn": "/sign-in",
    "signUp": "/sign-up",
  },
}
`;
